Framework/Geometry/src/Objects/Object.cpp

```diff
@@ -1475,6 +1482,153 @@ double Object::ConeSolidAngle(const V3D &observer,
+      accumulate;
+  // For seeding the single shot runs.
+  boost::random::ranlux48 rnEngine;
+  // Warm up statistics.
+  for (int i = 0; i < 10; ++i) {
+    const auto seed = rnEngine();
+    const double volume =
+        singleShotMonteCarloVolume(singleShotIterations, seed);
+    accumulate(volume);
+  }
+  const double relativeErrorTolerance = 1e-3;
+  double currentMean;
+  double currentError;
+  do {
+    const auto seed = rnEngine();
+    const double volume =
+        singleShotMonteCarloVolume(singleShotIterations, seed);
+    accumulate(volume);
+    currentMean = mean(accumulate);
+    currentError = error_of<tag::mean>(accumulate);
+    if (std::isnan(currentError)) {
+      currentError = 0;
+    }
+  } while (currentError / currentMean > relativeErrorTolerance);
+  return currentMean;
+}
+
+/**
+ * Calculates the volume using the Monte Carlo method.
+ * @param shotSize Number of iterations.
+ * @param seed A number to seed the random number generator.
+ * @returns The simulated volume of this object.
+ */
+double Object::singleShotMonteCarloVolume(const int shotSize,
+                                          const size_t seed) const {
+  const auto &boundingBox = getBoundingBox();
+  if (boundingBox.isNull()) {
+    throw std::runtime_error("Cannot calculate volume: invalid bounding box.");
+  }
+  int totalHits = 0;
+  const double boundingDx = boundingBox.xMax() - boundingBox.xMin();
+  const double boundingDy = boundingBox.yMax() - boundingBox.yMin();
+  const double boundingDz = boundingBox.zMax() - boundingBox.zMin();
+  PARALLEL {
+    const auto threadCount = PARALLEL_NUMBER_OF_THREADS;
+    const auto currentThreadNum = PARALLEL_THREAD_NUMBER;
+    size_t blocksize = shotSize / threadCount;
+    if (currentThreadNum == threadCount - 1) {
+      // Last thread may have to do threadCount extra iterations in
+      // the worst case.
+      blocksize = shotSize - (threadCount - 1) * blocksize;
+    }
+    boost::random::mt19937 rnEngine(
+        static_cast<boost::random::mt19937::result_type>(seed));
+    // All threads init their engine with the same seed.
+    // We discard the random numbers used by the other threads.
+    // This ensures reproducible results independent of the number
+    // of threads.
+    // We need three random numbers for each iteration.
+    rnEngine.discard(currentThreadNum * 3 * blocksize);
+    boost::random::uniform_01<double> rnDistribution;
+    int hits = 0;
+    for (int i = 0; i < static_cast<int>(blocksize); ++i) {
+      double rnd = rnDistribution(rnEngine);
+      const double x = boundingBox.xMin() + rnd * boundingDx;
+      rnd = rnDistribution(rnEngine);
+      const double y = boundingBox.yMin() + rnd * boundingDy;
+      rnd = rnDistribution(rnEngine);
+      const double z = boundingBox.zMin() + rnd * boundingDz;
+      if (isValid(V3D(x, y, z))) {
+        ++hits;
+      }
+    }
+    // Collect results.
+    PARALLEL_ATOMIC
+    totalHits += hits;
```

Comment:
@mantidproject/tech-steering Do we have a preference between std::atomic and openmp pragmas?